### PR TITLE
Fix int8 overflow in vectorize tests

### PR DESCRIPTION
**Pull request by asi1024**
Rel. https://github.com/cupy/cupy/pull/7338

### CODE REVIEW

File: tests/cupy_tests/functional_tests/test_vectorize.py [hunk 1]
```diff
@@ -85,7 +84,8 @@ def run_div(self, func, xp, dtypes):
         return f(x1, x2)
 
     @testing.for_all_dtypes_combination(names=('dtype1', 'dtype2'))
-    @testing.numpy_cupy_allclose(rtol=1e-6)
+    # TODO(asi1024): Fix return type.
+    @testing.numpy_cupy_allclose(rtol=1e-3, type_check=False)
```
Comment (asi1024) [inline tests/cupy_tests/functional_tests/test_vectorize.py:88]:
```py
>>> numpy.vectorize(lambda x, y: x / y)(numpy.array([1], dtype=numpy.uint16), numpy.array([1], dtype=numpy.float32)).dtype
dtype('float64')
```

Comment (asi1024) [reply tests/cupy_tests/functional_tests/test_vectorize.py:88]: @seberg Question: I found that the dtype of the result of true_divide of int16 scalar and float32 scalar in numpy is different between NumPy 1.24 and earlier. Is this an intentional change or a bug?

NumPy 1.23.5
```py
>>> numpy.__version__
'1.23.5'
>>> (numpy.int16(1) + numpy.float32(1)).dtype
dtype('float32')
>>> (numpy.int16(1) / numpy.float32(1)).dtype
dtype('float32')
```

NumPy 1.24.1
```py
>>> numpy.__version__
'1.24.1'
>>> (numpy.int16(1) + numpy.float32(1)).dtype
dtype('float32')
>>> (numpy.int16(1) / numpy.float32(1)).dtype
dtype('float64')
```

Comment (seberg) [reply tests/cupy_tests/functional_tests/test_vectorize.py:88]: Arrg, no that is a subtle bug in my new scalar logic :(.  The `int / int -> float64`, so it tries converting float32 to that and succeeds, but since the other is a float64, it must not do that.

Comment (jakirkham) [reply tests/cupy_tests/functional_tests/test_vectorize.py:88]: xref: https://github.com/numpy/numpy/pull/23079

cc @leofang (for awareness)

Comment (kmaehashi) [reply tests/cupy_tests/functional_tests/test_vectorize.py:88]: Thanks Sebastian for fixing this quickly!

@asi1024 Let's mark the test skipif numpy==1.24.0 or 1.24.1.

File: tests/cupy_tests/functional_tests/test_vectorize.py [hunk 2]
```diff
@@ -35,13 +34,13 @@ def my_add(x, y):
 
         return self._run(my_add, xp, [dtype1, dtype2])
 
-    @testing.for_all_dtypes_combination(names=('dtype1', 'dtype2'))
-    @testing.numpy_cupy_array_equal(accept_error=TypeError)
-    def test_vectorize_sub(self, xp, dtype1, dtype2):
+    @testing.for_dtypes('bhilqefdFD')
```
Comment (kmaehashi) [inline tests/cupy_tests/functional_tests/test_vectorize.py:37]: Talked with @asi1024 offline, this shrinks the test axis (combination of different dtypes) but it is acceptable as it is tested in `test_vectorize_add`.

File: tests/cupy_tests/functional_tests/test_vectorize.py [hunk 3]
```diff
@@ -624,20 +622,17 @@ def my_func(x1, x2):
...
         def my_func(x1, x2, x3):
             y = x1 + x2 * x3 ** x1
             x2 = y + x3 * x1
             return x1 + x2 + x3
 
         f = xp.vectorize(my_func)
-        x1 = testing.shaped_random((20, 30), xp, dtype, seed=1, scale=5)
-        x2 = testing.shaped_random((20, 30), xp, dtype, seed=2)
-        x3 = testing.shaped_random((20, 30), xp, dtype, seed=3)
+        x1 = testing.shaped_random((20, 30), xp, xp.float32, seed=1, scale=5)
```
Comment (kmaehashi) [inline tests/cupy_tests/functional_tests/test_vectorize.py:633]: Question: why is this change needed?

**Review by kmaehashi — DISMISSED** [review]
LGTM except for one question

File: tests/cupy_tests/functional_tests/test_vectorize.py (hunk 3 above)
Comment (asi1024) [reply tests/cupy_tests/functional_tests/test_vectorize.py:633]: Because this test overflows on int8 inputs.

Comment (kmaehashi) [reply tests/cupy_tests/functional_tests/test_vectorize.py:633]: I see, then how about shrinking the data size to, say, `(12, 16)`? As `TestVectorize` is a collection of realistic use-case tests, it makes sense to me to test dtype combination here.

Comment (asi1024) [reply tests/cupy_tests/functional_tests/test_vectorize.py]: Fixed in a1c8eac.

Comment (kmaehashi) [reply tests/cupy_tests/functional_tests/test_vectorize.py]: Thanks! (I just noticed I misread `shaped_random` as `shaped_arange` 😅 the fix looks good anyways)

**Review by kmaehashi — APPROVED** [review]
